### PR TITLE
feat: sampling-first task label generator for pre-training (#33)

### PR DESCRIPTION
**Pull request by mmcdermott**
## Summary

Adds `every_query.sample_tasks`, a three-stage Hydra script that replaces the exhaustive per-(shard × duration × code) label generation in `tasks.py` with a stochastic sampler drawing to a per-worker budget. Architecture, rationale, and design discussion are in issue #33.

This PR **adds** the sampler alongside existing `tasks.py` / `train.py::collate_tasks`; it does not delete them. Switching the training entry point is a follow-up once the sampler has been validated in a real PT run.

## Design highlights

- **Pure primitives**: `sample_tasks`, `sample_contexts`, `build_index_df`, `compute_max_time_per_subject`, `evaluate_index_df` — all side-effect-free and individually unit-tested.
- **Single-pass `evaluate_index_df`**: one `join_asof(strategy='forward', by=['subject_id','query'])` across the full index_df regardless of how many distinct codes are present, with a +1µs left-key shift to turn `>=` into strict `>`. Censoring is O(1) per row via a pre-computed `max_time` groupby.
- **Bit-identical to reference**: parametrized cross-check of `evaluate_index_df` against `tasks_reference.build_task_label_matrix` on the same synthetic fixture used by `tests/test_tasks.py`, across four durations.
- **Three idempotent stages** inside a single `run_worker` call: sampled-task JSON, unlabeled index parquet, and the final labeled task parquet. Each is written on first run and loaded (without resampling) on re-run unless `overwrite=true`. Pre-seeding any of the upstream artifacts is honored — the sampler uses them verbatim, which makes the script easy to debug by hand.
- **No parallelism code**: the script is a pure `(input_shard, task_shard, seed)` → output function. A sweep is literally `python -m every_query.sample_tasks -m input_shard=... task_shard=range(K)`. All parallelism is via Hydra launchers + multirun (SLURM array, Ray, joblib, …).
- **Seed derivation via blake2b** (cross-process stable — Python's builtin `hash` is not). `tasks_seed` depends only on `(seed, task_shard)` so varying `input_shard` within a fixed `task_shard` evaluates the **same** tasks on **different** patients — the property that makes a sweep produce a coherent PT dataset. `contexts_seed` depends on both axes.
- **Schema-compatible output**: the final parquet has exactly the columns `EveryQueryPytorchDataset` reads (`subject_id`, `prediction_time`, `boolean_value`, `occurs`, `query`, `duration_days`), so it drops into the existing dataloader without touching it.

## Tests

`tests/test_sample_tasks.py`: **29 unit tests + 2 doctests**, structured as four classes:

- `TestPrimitives` — schema, bounds, determinism, invalid-input rejection, `n=0` / empty-candidate edge cases for each primitive. Catches a subtle polars gotcha: `.unique()` is hash-based and order-unstable, which would leak into `.sample(seed=...)` non-determinism without an explicit sort — now tested.
- `TestEvaluateAgainstReference` — parametrized bit-identity cross-check against `tasks_reference.build_task_label_matrix` across four durations, plus targeted edge-case probes for codes-not-in-shard and for the strict-`>` boundary at `prediction_time == event_time`.
- `TestRunWorkerPipeline` — artifact-level tests over a synthetic fake-cohort tmpdir:
    - all three artifacts written on first run,
    - idempotent rerun is a no-op and leaves bytes untouched,
    - `overwrite=True` regenerates,
    - `task_shard` axis changes tasks,
    - `input_shard` axis preserves tasks (the key property for coherent sweeps),
    - pre-seeded `tasks.json` on disk is honored verbatim.
- `TestEndToEndWithDataset` — hand-builds a sampler-shaped labels parquet referencing real subject IDs from `tensorized_cohort_dir`, instantiates `EveryQueryPytorchDataset` on it + the real tokenized shards, and pushes a collated batch through `demo_model`. Verifies schema compatibility with the downstream dataloader without depending on the early-stage string-coded events parquet layout (which the test fixture does not expose — only the integer-coded `normalization/` stage is materialized).

Full suite: **207 passed** locally (176 pre-existing + 29 new unit + 2 new doctests). Pre-commit clean on all touched files.

## Explicitly out of scope (per issue #33 discussion)

- **Composite task grammar** (`AnyTask`, `AllTask`, conjunctions) — separate follow-up PR. The only function that needs to change is `evaluate_index_df`.
- **Deleting `tasks.py` / `train.py::collate_tasks`** — will coexist until the sampler is validated end-to-end in a real PT run.
- **MEDS-Transforms wrapping** — only if there's a concrete reuse benefit.
- **Switching `train.py` to consume sampler output** — follow-up PR so this one stays focused.

## Test plan

- [x] `uv run pytest tests/test_sample_tasks.py -q` — 29 passed.
- [x] `uv run pytest --doctest-modules src/every_query/sample_tasks.py -q` — 2 passed.
- [x] `uv run pytest -q` (full suite) — 207 passed.
- [x] `uv run pre-commit run --files src/every_query/sample_tasks.py src/every_query/sample_tasks_config.yaml tests/test_sample_tasks.py` — all 23 hooks pass.
- [ ] Smoke-run on a real cohort with `hydra -m input_shard=0,1 task_shard=0,1` to confirm parallelism works end-to-end.
- [ ] Spot-check a labels parquet manually to confirm the balance of `occurs` / `boolean_value` is sane.

Closes part of #33 (the sampler; composite grammar and `tasks.py` cleanup are follow-ups).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: tests/test_sample_tasks.py [hunk 1]
```diff
@@ -0,0 +1,622 @@
...
+            "min_context_per_subject": 5,
+        }
+        first = run_worker(seed=1, **kwargs)
+        second = run_worker(seed=2, overwrite=True, **kwargs)
+        # Different seeds + overwrite → different labels file (tasks drawn differently).
+        assert first is not None and second is not None
+        assert first == second  # same path
+        # The underlying bytes should change because the tasks re-sampled.
+        first_tasks = (out_dir / "_artifacts" / "tasks" / "train" / "0__0000.json").read_text()
+        assert first_tasks is not None  # tasks file was regenerated
```
[the remark this answers is not in the document]
Comment (mmcdermott) [reply tests/test_sample_tasks.py]: Fixed in 6ac4824. Applied your suggested test rewrite verbatim: the rerun now captures `first.read_bytes()` and `tasks_path.read_bytes()`, asserts they're unchanged after a no-op `run_worker(seed=1, ...)` (second call returns `None`), then asserts both change after `run_worker(seed=2, overwrite=True, ...)`. Catches both directions of the bug (overwrite silently ignored, overwrite=False silently regenerating).

File: src/every_query/sample_tasks.py [hunk 2]
```diff
@@ -0,0 +1,516 @@
...
+    split: str,
+    input_shard: str,
+    task_shard: int,
+) -> tuple[Path, Path, Path]:
+    """Resolve the ``(tasks_fp, index_fp, labels_fp)`` triple for one worker."""
+    worker_id = f"{input_shard}__{task_shard:04d}"
+    tasks_fp = out_dir / "_artifacts" / "tasks" / split / f"{worker_id}.json"
+    index_fp = out_dir / "_artifacts" / "index" / split / f"{worker_id}.parquet"
+    labels_fp = out_dir / split / f"{worker_id}.parquet"
+    return tasks_fp, index_fp, labels_fp
```
[the remark this answers is not in the document]
Comment (mmcdermott) [reply src/every_query/sample_tasks.py]: Good catch — this was a real silent-staleness hazard. Fixed in 6ac4824 by adding a per-worker meta sidecar at `_artifacts/runs/{split}/{worker_id}.json` that records the full sampling config (seed, n_tasks, contexts_per_task, duration_min, duration_max, min_context_per_subject). `run_worker` validates the on-disk meta against the requested config on every non-overwrite rerun and raises a `ValueError` with a per-key diff on mismatch:

```
Refusing to reuse cached sampler artifacts: the worker config on disk at
.../_artifacts/runs/train/0__0000.json differs from the requested config:
  n_tasks: on_disk=16, current=32
Pass overwrite=true to regenerate this worker's artifacts, or delete the
_artifacts/ subdirectory if you want a clean run.
```

Absent meta is tolerated — legacy artifacts from before this change and hand-preseeded `tasks.json` files (used for debugging per issue #33) keep working.

Three new tests pin the behavior:
- `test_config_mismatch_raises_without_overwrite`: changing `n_tasks` raises, `overwrite=True` bypasses and regenerates correctly (verified by row count).
- `test_config_mismatch_raises_on_seed_change`: changing `seed` raises with an error mentioning `seed`.
- `test_missing_meta_is_tolerated`: deleting the meta file leaves the legacy / preseeded path working.

Also went through the module for other issues while I was in there and found two:
- `write_parquet` / `write_text` wrote directly to the target path; a crash mid-write would leave a corrupt file the next rerun would happily load. Added `_atomic_write_text` / `_atomic_write_parquet` helpers that write to a `.tmp` sibling and rename.
- `test_input_shard_axis_preserves_tasks` only checked tasks were identical across input shards — it didn't verify contexts actually differ, which is the property that makes cross-shard parallelism useful. Extended to compare the `(subject_id, prediction_time)` multiset of each index file.

File: src/every_query/sample_tasks.py [hunk 3]
```diff
@@ -0,0 +1,616 @@
...
+def _read_event_shard(file_path: str | Path) -> pl.DataFrame:
+    """Read a shard parquet and return events as ``(subject_id, time, code)`` sorted by subject+time.
+
+    Unlike ``tasks.read_event_shard``, this keeps ``code`` as ``Utf8`` rather than casting to
+    ``Categorical`` — the sampler joins events against ``query`` (which is a plain Utf8 string column
+    in the index_df), and a mixed Categorical/Utf8 join would error.
+    """
+    return (
+        pl.read_parquet(file_path)
+        .select(["subject_id", "time", "code"])
```
[the remark this answers is not in the document]
Comment (mmcdermott) [reply src/every_query/sample_tasks.py]: Fixed in cd2332f. `_read_event_shard` now explicitly casts `code` → `pl.Utf8` and `time` → `pl.Datetime("us")` on read. You're right that both matter:

- The fixture's `normalization/train/0.parquet` stage stores `code: UInt8` (vocab indices), `tasks.read_event_shard` casts to `Categorical`, and a plain export might be `Utf8`. Joining any of those against the `Utf8` `query` column in `index_df` either errors or silently produces zero matches.
- The `+1µs` shift only does its job at microsecond precision. At ms it rounds to zero and silently turns `>` into `>=` — a latent correctness bug for any shard whose parquet was written at ms precision.

Added `TestReadEventShardDtypeNormalization` with three cases: Categorical `code` → Utf8 round-trip, ms `time` → us round-trip, and an end-to-end test that feeds a Categorical-coded shard through `_read_event_shard` + `evaluate_index_df` and asserts the labels are correct (i.e. the join actually matches, rather than silently producing zero rows).

File: src/every_query/sample_tasks.py [hunk 4]
```diff
@@ -0,0 +1,616 @@
...
+    tmp.write_text(text)
+    tmp.replace(fp)
+
+
+def _atomic_write_parquet(df: pl.DataFrame, fp: Path) -> None:
+    """Write ``df`` to ``fp`` atomically via tmpfile + rename."""
+    fp.parent.mkdir(parents=True, exist_ok=True)
+    tmp = fp.with_suffix(fp.suffix + ".tmp")
+    df.write_parquet(tmp)
+    tmp.replace(fp)
```
[the remark this answers is not in the document]
Comment (mmcdermott) [reply src/every_query/sample_tasks.py]: Fixed in cd2332f. Added `_unique_tmp_path` that uses `tempfile.mkstemp(prefix=f".{fp.name}.", suffix=".tmp", dir=fp.parent)` to allocate a process-unique sibling file per call, then routed both `_atomic_write_text` and `_atomic_write_parquet` through it and switched the rename to `os.replace`. Wrapped the write in `try/except` + `Path.unlink(missing_ok=True)` so a mid-write failure cleans up after itself instead of leaving an orphan tmp sitting next to the target.

Added `TestAtomicWriteConcurrency` with three assertions:
- Two consecutive `_unique_tmp_path` calls for the same target return distinct names in the target's parent directory.
- `_atomic_write_text` round-trips and leaves zero orphan files after a successful write.
- Same for `_atomic_write_parquet`.

A true race-condition test would need real concurrency and timing control, which is flaky to write — but the "different names per call" property is what makes the race impossible to begin with, so that's the invariant the test pins.

File: src/every_query/sample_tasks.py [hunk 5]
```diff
@@ -0,0 +1,616 @@
...
+    window_end = pl.col("prediction_time") + duration_expr
+    censored = window_end > pl.col("max_time")
+    event_in_window = pl.col("time").is_not_null() & (pl.col("time") < window_end)
+
+    return (
+        joined.with_columns(censored.alias("_censored"))
+        .with_columns(
+            pl.col("_censored").alias("boolean_value"),
+            (pl.col("_censored").not_() & event_in_window).alias("occurs"),
+        )
```
[the remark this answers is not in the document]
Comment (mmcdermott) [reply src/every_query/sample_tasks.py:357]: Fixed in cd2332f. Went with policy (b): treat missing-subject rows as censored (`boolean_value=True`, `occurs=False`) and emit a warning via `logger.warning(...)` so the condition is visible.

Rationale for (b) over raising:
- The standard flow (`sample_contexts` → `build_index_df` → `evaluate_index_df`) always draws contexts from the same `events_df` the evaluator uses, so missing-subject rows only happen when someone pre-seeds or hand-edits an `index_df` referencing subjects outside the current shard — the "missing data = censored" semantic is the most intuitive default for that edge case.
- It matches the existing censoring semantic ("we don't know if the event happened within the window") rather than introducing a new failure mode.
- A warning makes the condition visible without breaking the pipeline, so a user who has legitimately pre-seeded cross-shard subjects still gets a run and a clear signal.

Implemented as `(window_end > pl.col("max_time")).fill_null(True)` on the `censored` expression, which combined with `(~censored & event_in_window)` for `occurs` propagates the "censored" default cleanly through both output columns.

New test `test_unknown_subject_is_treated_as_censored` asserts: no null booleans escape the evaluator, unknown-subject rows come out as censored/not-occurred, and the warning is emitted (caught via `caplog.at_level("WARNING", logger="every_query.sample_tasks")`).

File: src/every_query/sample_tasks_config.yaml [hunk 6]
```diff
@@ -0,0 +1,34 @@
+# Hydra config for the sampling-first task label generator.
+#
+# Sweep on the cluster with:
+#   python -m every_query.sample_tasks -m input_shard=0,1,2,... task_shard=range(0,K)
+#
+# Defaults pull data paths from the same env vars as tasks.py (see _env.py) so CLI users don't
+# have to repeat them, but tests can inject values directly with `data_dir=/path out_dir=/path`.
+
+data_dir: ${oc.env:INTERMEDIATE,null} # root containing data/{split}/*.parquet and metadata/codes.parquet
```
[the remark this answers is not in the document]
Comment (mmcdermott) [reply src/every_query/sample_tasks_config.yaml]: Fixed in daab561. You're right — I verified against `.env.example` that `$INTERMEDIATE` and `$PROCESSED` are distinct siblings of `$DATA_DIR`, and `tasks.py:215-217` confirms the sampler was conflating them.

The config yaml now has three separate `data_dir` / `codes_dir` / `out_dir` entries, all defaulting to `null`:

```yaml
data_dir: null
codes_dir: null
out_dir: null
```

The `${oc.env:...}` interpolations are gone entirely — see the related comment about `load_dotenv()`; they resolved too early for the repo's dotenv convention, so the fallback is now done inside `main()` via a new `_resolve_path` helper. Comment block at the top of the yaml documents the fallback mapping explicitly.

File: src/every_query/sample_tasks.py [hunk 7]
```diff
@@ -0,0 +1,672 @@
...
+
+def _read_query_codes(data_dir: str | Path) -> list[str]:
+    """Read the universe of query codes from ``{data_dir}/metadata/codes.parquet``.
+
+    The ``.unique()`` result is explicitly sorted because polars' default hash-based unique is
+    order-unstable across distinct DataFrame instances — two calls in the same Python session
+    can return the codes in different orders, which would make ``sample_tasks`` non-deterministic
+    with respect to the tasks seed across workers reading the same metadata file.
+    """
+    codes_fp = Path(data_dir) / "metadata" / "codes.parquet"
```
[the remark this answers is not in the document]
Comment (mmcdermott) [reply src/every_query/sample_tasks.py]: Fixed in daab561. Threaded an explicit `codes_dir` parameter through `_read_query_codes`, `run_worker`, and `main()`. Signature is now:

```python
def _read_query_codes(codes_dir: str | Path) -> list[str]:
    codes_fp = Path(codes_dir) / "metadata" / "codes.parquet"
    ...
```

and `run_worker` takes `codes_dir` as a required keyword. `main()` resolves `codes_dir` via the same `_resolve_path` helper used for `data_dir` / `out_dir`, defaulting to `$PROCESSED` as a fallback — matching `tasks.py:215`'s `read_codes_dir = os.environ["PROCESSED"]`.

Also updated `_write_fake_cohort` to write codes.parquet under a sibling `{tmp}/processed/` rather than under the shard root, so the test cohort actually mirrors the `.env.example` layout:

```
{tmp}/intermediate/data/{split}/*.parquet    (data_dir)
{tmp}/processed/metadata/codes.parquet       (codes_dir)
```

Added `TestRunWorkerPipeline::test_codes_dir_is_read_from_a_distinct_root` as a regression: asserts codes.parquet is *absent* under `data_dir` and *present* under `codes_dir`, then runs the worker and checks the output queries. Without the fix this test raises `FileNotFoundError` at `_read_query_codes` time.

File: src/every_query/sample_tasks.py [hunk 8]
```diff
@@ -0,0 +1,672 @@
...
+@hydra.main(version_base=None, config_path=".", config_name="sample_tasks_config")
+def main(cfg: DictConfig) -> None:
+    """Hydra entry point.
+
+    See :func:`run_worker` for the per-worker pipeline.
+    """
+    if cfg.data_dir is None:
+        raise ValueError("data_dir must be set (pass data_dir=... or export $INTERMEDIATE)")
+    if cfg.out_dir is None:
+        raise ValueError("out_dir must be set (pass out_dir=... or export $TASK_DIR)")
```
[the remark this answers is not in the document]
Comment (mmcdermott) [reply src/every_query/sample_tasks.py]: Fixed in daab561. `main()` now calls `load_dotenv()` immediately before resolving the three path roots, matching `_env.py`/`tasks.py`. Important subtlety: I also **removed the `${oc.env:...}` interpolations from the yaml**, because Hydra resolves those when it builds the config — *before* `main()` has a chance to call `load_dotenv()`. Keeping both would mean "env var set in `.env` but not exported" still resolves to `null`, which is exactly the bug you flagged.

The new flow:
1. `sample_tasks_config.yaml` defaults `data_dir` / `codes_dir` / `out_dir` to `null`.
2. `main()` calls `load_dotenv()` (late-imported so it doesn't run at module import).
3. `main()` calls `_resolve_path(cfg.get("data_dir"), "INTERMEDIATE", "data_dir")` and the two siblings. Explicit CLI value wins over env var wins over raise-with-clear-message.

Error message includes a `.env` hint so users who hit it get pointed at the right mechanism:

```
data_dir must be set: pass data_dir=... on the CLI, set it in sample_tasks_config.yaml,
or export $INTERMEDIATE (or define it in .env — sample_tasks calls load_dotenv()).
```

`TestResolvePath` covers the full fallback matrix (explicit wins / env fallback / both unset raises / error mentions env var + `.env` / empty env var treated as unset) via `monkeypatch.setenv`. And I went *lighter* than `ensure_env()` here intentionally — it requires all seven `REQUIRED_ENV_VARS` even though the sampler only uses three, and `WANDB_ENTITY` / `PROJECT_DIR` aren't relevant to this script.

File: src/every_query/sample_tasks.py [hunk 9]
```diff
@@ -0,0 +1,708 @@
...
+        - ``occurs = (not censored) AND (next event with matching code falls strictly within
+          (prediction_time, prediction_time + duration_days))``
+        - ``boolean_value = censored`` (alias expected by ``EveryQueryPytorchDataset``)
+
+    The ``>`` on event time is enforced by shifting the asof key by ``+1µs`` since datetimes are
+    stored at microsecond precision, which turns ``strategy="forward"``'s ``>=`` into a strict ``>``.
+
+    Args:
+        index_df: Output of ``build_index_df``. Must have columns ``subject_id``, ``prediction_time``,
+            ``query``, ``duration_days`` (``task_id`` is optional and preserved if present).
```
[the remark this answers is not in the document]
Comment (mmcdermott) [reply src/every_query/sample_tasks.py]: Fixed in 15898e7. Good catch — the drop has always been intentional (there's already a code comment in `run_worker` saying "Downstream MEDS dataloader does not use task_id; it is intentionally absent from the published schema"), only the Args docstring was stale. Applied your suggested wording on the ``index_df`` arg and extended it to call out *why* the schema is limited — it's pinned to what ``EveryQueryPytorchDataset`` reads, not an accident of implementation.

File: src/every_query/sample_tasks.py [hunk 10]
```diff
@@ -0,0 +1,710 @@
...
+    code: str
+    duration_days: int
+
+
+# ---------------------------------------------------------------------------
+# Seed derivation
+# ---------------------------------------------------------------------------
+
+
+def derive_seed(*parts: int | str) -> int:
```
Comment (payalchandak) [inline src/every_query/sample_tasks.py:59]: This should live in utils/ and be standardized across the repo, created issue for future PR #52

**Review by payalchandak — APPROVED** [review]

**Comment by payalchandak** [comment]
Need to refactor out `tasks.py` and other dead code https://github.com/payalchandak/EveryQuery/issues/53
